### ISSUE TITLE
Provide an API to register inline scripts or styles

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor#2539]
Currently when we want to add a contextual script or style in an extension, we need to add it with the content:

```ruby
  def process parent, target, attrs
    name = attrs['name']
    html = %(
  <div class="block">Say my name</div>
  <script type="text/javascript">console.log('Hello #{name}');</script>
)
    create_pass_block parent, html, attrs, subs: nil
  end
```

As far as I know it's not possible to register a `<script>` or a `<style>` to be appended elsewhere in the HTML5 document.

It could be useful to be able to register a script or a style:

```ruby
Extensions::HTMLProcessor.register_script(script_content)
```

By default I think we should append `<script>` just before the end of the `<body>` (or in the `<head>` with the `defer` attribute) and we should append inline `<style>` in the `<head>`.

This API will be useful in the browser extension because currently we are forced to use the `html()` function from JQuery (to `eval` inline `<script>`).
This function is unsafe and most likely significantly slower than `innerHTML`.

**Comment by mojavelinux** [comment]
I certainly want Asciidoctor to offer this capability. Is this not something the DocinfoProcessor can provide? If you need more control, we could consider more insertion slots.

**Comment by Mogztter** [comment]
> I certainly want Asciidoctor to offer this capability. Is this not something the DocinfoProcessor can provide? 

I don't think so because the JavaScript code is generated from the context of an extension. For instance, if I have the following definition in my document:

```
[chart,line]
....
January,February,March
28,48,40
65,59,80
....
```

The generated code will look like:

```js
var options = {
  height: '400',
  width: '600',
  colors:["#72B3CC", "#8EB33B"]
};
var data = {
  labels: ["January", "February", "March"],
  series: [["28", "48", "40"], ["65", "59", "80"]]
};
new Chartist.Line('#chartNHZHMPCW', data, options);
```

To work around this issue and use a `DocinfoProcessor`, we could define [custom data attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) and use a global script to initialize all the charts.

In other words, the chart block macro extension will no longer generate the JavaScript code but will generate the following `<div>`:

```html
<div id="chartXWTQLHLM" 
     class="chart" 
     data-chart-labels="January,February,March"
     data-chart-type="line"
     data-chart-series-0="28,48,40"
     data-chart-series-1="65,59,80">
</div>
```

And then the `DocinfoProcessor` can iterate on all the `div` with the `chart` class and initialize the chart.

> If you need more control, we could consider more insertion slots.

I think we need to find a mechanism to share data/context with the `DocinfoProcessor`.
